### PR TITLE
feat: Merkle Distributor Contracts

### PR DESCRIPTION
**Pull request by piesrtasty**
- Contracts for MerkleDistributor and MerkleDistributorFactory
- Tests for MerkleDistributor and MerkleDistributorFactory

### CODE REVIEW

**Review by wei3erHase — COMMENTED** [review]
looks clean af! 👌 
i'd move `MerkleDistributor.sol` to either `/tokens` directory (where the TokenDistributor is) or to `/governance` (related). `/utils` directory is for abstract inheritable contracts in general.
confirming that comparting MerkleDistributor to TokenDistributor only `mint > transfer` relevant change.

**Comment by piesrtasty** [comment]
> looks clean af! 👌 i'd move `MerkleDistributor.sol` to either `/tokens` directory (where the TokenDistributor is) or to `/governance` (related). `/utils` directory is for abstract inheritable contracts in general. confirming that comparting MerkleDistributor to TokenDistributor only `mint > transfer` relevant change.
> 
> given that the factory will be called directly through DAO UI, don't you prefer to specify each parameter instead of an encoded struct? will make your future life much easier

Great suggestions! Will implement these along w/ tests.

File: src/contracts/tokens/MerkleDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    return _canClaim(_proof, _user, _amount);
+  }
+
+  /// @inheritdoc IMerkleDistributor
+  function claim(bytes32[] calldata _proof, uint256 _amount) external {
+    _claim(_proof, _amount);
+  }
+
+  /// @inheritdoc IMerkleDistributor
+  // function sweep(address _sweepReceiver) external override isAuthorized {
```
Comment (wei3erHase) [inline src/contracts/tokens/MerkleDistributor.sol:69]: ?

File: test/mocks/MerkleDistributorFactoryForTest.sol
Comment (wei3erHase) [inline test/mocks/MerkleDistributorFactoryForTest.sol:1]: not needed

File: test/unit/tokens/MerkleDistributor.t.sol [hunk 2]
```diff
@@ -0,0 +1,480 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import 'forge-std/console.sol';
```
Comment (wei3erHase) [inline test/unit/tokens/MerkleDistributor.t.sol:4]: 👀

File: test/unit/tokens/MerkleDistributor.t.sol [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    _mockClaimed(distroRecipients[4], true);
+    assertFalse(merkleDistributor.canClaim(validEveProofs, distroRecipients[4], distroAmounts[4]));
+  }
+
+  function test_CannotClaimZeroAmount() public {
+    assertFalse(merkleDistributor.canClaim(validEveProofs, distroRecipients[4], 0));
+  }
+}
+
+contract Unit_CanClaim_ExternalScript is Base {
```
Comment (wei3erHase) [inline test/unit/tokens/MerkleDistributor.t.sol:220]: what's `ExternalScript`?

File: test/unit/tokens/MerkleDistributor.t.sol (hunk 2 above)
Comment (piesrtasty) [reply test/unit/tokens/MerkleDistributor.t.sol:4]: oh dear, need to remove that!

**Pull request closed without merging**